#' Compare Gene Expression by Peak Accessibility
#' Compares normalised gene expression between cells with and without accessibility at a linked peak. Includes total RNA UMI counts to check for the impact of library size. Returns a tibble by default or a ggplot object if \code{plot = TRUE}.
#'
#' @param links_data A tibble generated by \code{get_link_data()}
#' @param seurat_object A Seurat object containing peak-to-gene links generated by \code{Signac::LinkPeaks()}
#' @param plot Logical; if \code{TRUE}, returns a ggplot object visualising expression and UMI distributions. Default is \code{FALSE}
#'
#' @returns A tibble with one row per cell-peak-gene combination and the following columns:
#' \itemize{
#'    \item \code{gene}: the selected linked gene
#'    \item \code{peak}: the selected linked peak
#'    \item \code{score}: the correlation coefficient - from LinkPeaks
#'    \item \code{zscore}: the z-score of the correlation coefficient - from LinkPeaks
#'    \item \code{pvalue}: the p-value associated with the z-score - from LinkPeaks
#'    \item \code{cell_barcode}: the unique identifier for each cell - from Seurat Object
#'    \item \code{gene_data}: the normalised RNAseq expression data - from Seurat Object
#'    \item \code{gene_counts}: the raw RNAseq count data - from Seurat Object
#'    \item \code{peak_data}: the normalised ATAC expression data - from Seurat Object
#'    \item \code{peak_counts}: the raw ATAC count data - from Seurat Object
#'    \item \code{read_in_peak}: Logical; \code{TRUE} if a read is observed in the peak for a given cell
#'    \item \code{nCount_RNA}: Total RNA UMI counts per cell - from Seurat Object
#'
#' @export
#'
#' @examples
by_peak_expression <- function(links_data, seurat_object,plot = FALSE){

  # Get the total UMI counts from the meta data
  tibble::as_tibble(seurat_object@meta.data,rownames = "cell_barcode") %>%
    dplyr::select(cell_barcode,nCount_RNA) %>%
    dplyr::right_join(links_data,by = "cell_barcode") %>%
    dplyr::relocate(cell_barcode,.after = pvalue) %>%
    dplyr::relocate(nCount_RNA, .after = read_in_peak) -> links_data


  if (plot == FALSE){
    return(links_data)
  }

  else{

    # add a label name for plots
    links_data %>%
      tidyr::pivot_longer(cols = c(gene_data, nCount_RNA), names_to = "count_type", values_to = "value") %>%
      dplyr::mutate(count_type = stringr::str_replace(count_type,"gene_data","Normalised Gene Expression"),
                    count_type = stringr::str_replace(count_type,"nCount_RNA","Total UMI Count")
                    ) %>%
      dplyr::mutate(label_name = stringr::str_c(gene, peak, sep = " : ")) %>%
      dplyr::mutate(label_name = stringr::str_c(label_name, round(score, 3), sep = "\n")) %>%
      dplyr::group_by(count_type) %>%
      dplyr::group_split() %>%
      purrr::map(~ {
        ggplot2::ggplot(.x, ggplot2::aes(x = read_in_peak, y = value, colour = read_in_peak)) +
          ggplot2::geom_jitter(width = 0.3) +
          ggplot2::geom_violin(colour = "black", fill = NA) +
          ggplot2::facet_wrap(~label_name, ncol = 1, scales = "free") +
          ggplot2::ggtitle(unique(.x$count_type)) +
          ggplot2::xlab(NULL) +
          ggplot2::ylab(NULL) +
          ggplot2::theme(
            plot.title = ggplot2::element_text(hjust = 0.5, size = 12),
            legend.position = "top",
            legend.text = ggplot2::element_text(size = 10),
            strip.text = ggplot2::element_text(size = 10),
            strip.background = ggplot2::element_rect("white"),
            axis.text.x = ggplot2::element_blank(),
            axis.ticks.x = ggplot2::element_blank()
          )
      }) %>%
      patchwork::wrap_plots(ncol = 2) +
      patchwork::plot_annotation(
        theme = ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5))
      )
  }

}
