#' Get Data Associated with Peak-to-Gene Links
#' Extracts data related to specified peak-to-gene links from a Seurat object, including gene expression and peak accessibility for each cell, and determines whether a read is present in the linked peak per cell.
#'
#'
#' @param seurat_object A Seurat object containing peak-to-gene links generated by \code{Signac::LinkPeaks()}
#' @param Gene Character vector of gene name(s) of links to include
#' @param Peak Character vector of peak name(s) of links to include
#' @param Pvalue Numeric value specifying the maximum p-value threshold for links to include
#' @param expression.assay Name of the assay containing gene expression information
#' @param chromatin.assay Name of the assay containing peak information
#'
#' @returns A tibble with one row per cell-peak-gene combination and the following columns:
#' \itemize{
#'    \item \code{gene}: the selected linked gene
#'    \item \code{peak}: the selected linked peak
#'    \item \code{score}: the correlation coefficient - from LinkPeaks
#'    \item \code{zscore}: the z-score of the correlation coefficient - from LinkPeaks
#'    \item \code{pvalue}: the p-value associated with the z-score - from LinkPeaks
#'    \item \code{cell_barcode}: the unique identifier for each cell - from Seurat Object
#'    \item \code{gene_data}: the normalised RNAseq expression data - from Seurat Object
#'    \item \code{gene_counts}: the raw RNAseq count data - from Seurat Object
#'    \item \code{peak_data}: the normalised ATAC expression data - from Seurat Object
#'    \item \code{peak_counts}: the raw ATAC count data - from Seurat Object
#'    \item \code{read_in_peak}: Logical; \code{TRUE} if a read is observed in the peak for a given cell
#' @export
#'
#' @examples get_link_data(seurat_object = data,Gene = c("MS4A1","LYZ"))
#'
get_link_data <- function(seurat_object, Gene=NULL, Peak=NULL,Pvalue=NULL,
                          expression.assay = "RNA", chromatin.assay = "ATAC"){

  # General checks on seurat object

  # has a seurat_object with link data been provided
  if (!base::inherits(seurat_object,"Seurat")) {
    base::stop(stringr::str_c(base::deparse(substitute(seurat_object)), "is not a Seurat object.", sep = " "))
  }

  if( !expression.assay %in% names(seurat_object@assays) ) {
    base::stop(stringr::str_c(expression.assay, "is not an assay in", base::deparse(base::substitute(seurat_object)), sep = " "))
  }

  if( !chromatin.assay %in% names(seurat_object@assays) ) {
    base::stop(stringr::str_c(chromatin.assay, "is not an assay in", base::deparse(base::substitute(seurat_object)), sep = " "))
  }


  # Extract links data as tibble
  tibble::as_tibble(Signac::Links(seurat_object)) -> selected_links

  if (! base::is.null(Gene)) {
    # Gene should be a character vector
    assertthat::assert_that(base::is.character(Gene))

    #check all genes are in the tibble
    if ( base::any(!Gene %in% dplyr::pull(selected_links,gene))) {
      warning("at least one of the genes provided is missing from the FeatureLink data")
    }

    selected_links %>%
      dplyr::filter(gene %in% Gene) -> selected_links
  }

  if (! base::is.null(Peak)){
    # Peak should be a character vector
    assertthat::assert_that(base::is.character(Peak))

    #check all genes are in the tibble
    if ( base::any(!Peak %in% dplyr::pull(selected_links,peak))) {
      warning("at least one of the peaks provided is missing from the FeatureLink data")
    }

    selected_links %>%
      dplyr::filter(peak %in% Peak) -> selected_links
  }

  if (! base::is.null(Pvalue)){
    # Pvalue if provided should be a single numeric value
    assertthat::assert_that(assertthat::is.number(Pvalue),msg = "Pvalue must be a single number")

    selected_links %>%
      dplyr::filter(pvalue < Pvalue) -> selected_links
  }

  selected_links %>%
    dplyr::select(gene,peak,score,zscore,pvalue) -> selected_links

  # Get the Raw and Normalised Data Associated with these links
  # might need something here to check for an empty tibbles/ assays

  # First get the Raw ATAC data
  SeuratObject::GetAssayData(seurat_object, assay = chromatin.assay, layer = "counts") -> atac_data
  selected_links %>% dplyr::distinct(peak) %>% dplyr::pull(peak) -> peaks
  atac_data[base::rownames(atac_data) %in% peaks, , drop = FALSE] -> atac_data

  tibble::as_tibble(atac_data,rownames = "peak") %>%
    dplyr::right_join(selected_links,by = "peak") %>%
    tidyr::pivot_longer(cols = !c(peak,gene,score,zscore,pvalue),
                        names_to = "cell_barcode",values_to = "peak_counts") -> selected_links_data

  # Next get the normalised ATAC data associated with these links

  SeuratObject::GetAssayData(seurat_object, assay = chromatin.assay, layer = "data") -> atac_data
  selected_links %>% dplyr::distinct(peak) %>% dplyr::pull(peak) -> peaks
  atac_data[base::rownames(atac_data) %in% peaks, , drop = FALSE] -> atac_data

  tibble::as_tibble(atac_data,rownames = "peak") %>%
    dplyr::right_join(selected_links,by = "peak") %>%
    dplyr::select(-c(score,zscore,pvalue)) %>%
    tidyr::pivot_longer(cols = !c(peak,gene),
                        names_to = "cell_barcode",values_to = "peak_data") %>%
    dplyr::right_join(selected_links_data, by = c("cell_barcode", "peak","gene")) -> selected_links_data


  # Next Get Raw RNAseq data

  SeuratObject::GetAssayData(seurat_object, assay = expression.assay, layer = "counts") -> gene_data
  selected_links %>% dplyr::distinct(gene) %>% dplyr::pull(gene) -> genes
  gene_data[base::rownames(gene_data) %in% genes, , drop = FALSE] -> gene_data

  tibble::as_tibble(gene_data,rownames= "gene") %>%
    dplyr::right_join(selected_links,by = "gene") %>%
    dplyr::select(-c(score,zscore,pvalue)) %>%
    dplyr::relocate("gene", .after = "peak") %>%
    tidyr::pivot_longer(cols = !c(peak,gene),names_to = "cell_barcode",values_to = "gene_counts") %>%
    dplyr::right_join(selected_links_data, by = c("cell_barcode", "peak","gene")) %>%
    dplyr::relocate(c(peak,score,zscore,pvalue),.after = "gene") -> selected_links_data


  # Finally get the normalised RNAseq data

  SeuratObject::GetAssayData(seurat_object, assay = expression.assay, layer = "data") -> gene_data
  selected_links %>% dplyr::distinct(gene) %>% dplyr::pull(gene) -> genes
  gene_data[base::rownames(gene_data) %in% genes, , drop = FALSE] -> gene_data

  tibble::as_tibble(gene_data,rownames= "gene") %>%
    dplyr::right_join(selected_links,by = "gene") %>%
    dplyr::select(-c(score,zscore,pvalue)) %>%
    dplyr::relocate("gene", .after = "peak") %>%
    tidyr::pivot_longer(cols = !c(peak,gene),names_to = "cell_barcode",values_to = "gene_data") %>%
    dplyr::right_join(selected_links_data, by = c("cell_barcode", "peak","gene")) %>%
    dplyr::relocate(c(peak,score,zscore,pvalue),.after = "gene") -> selected_links_data

  # Determine if there is a read in the peak
  selected_links_data %>%
    dplyr::mutate(read_in_peak = dplyr::if_else(peak_counts !=0,"yes","no")) -> selected_links_data


  return(selected_links_data)
}
